### PR TITLE
Vnext/775 inadequate budget reason

### PR DESCRIPTION
**Pull request by gregglarson**
This needs to be pulled into the Simulation for determining if analysis is running correctly.  Updates REASONS_ table creation and output.

### CODE REVIEW

File: iAM-legacy/RoadCare3/FormAnalysis.cs [hunk 1]
```diff
@@ -1599,5 +1607,27 @@ private void DataGridViewRemainLife_CellMouseDoubleClick(object sender, DataGrid
                 }
             }
         }
+
+        private void CheckBoxReasons_CheckedChanged(object sender, EventArgs e)
+        {
+            var update = "";
+            if (checkBoxReasons.Checked)
+            {
+                update = "UPDATE SIMULATIONS SET USE_REASONS='1' WHERE SIMULATIONID='" + m_strSimulationID + "'";
```
Comment (saranshbhardwaj) [inline iAM-legacy/RoadCare3/FormAnalysis.cs:1616]: If we are using USE_REASON as a boolean, shouldn't we change it's type in the database?

Comment (ARA-Trans) [reply iAM-legacy/RoadCare3/FormAnalysis.cs:1616]: I believe this is a 1 or a 0 as the MSSQL database uses a "bit" field to handle true/false items. I don't think bool exists there.

Comment (gregglarson) [reply iAM-legacy/RoadCare3/FormAnalysis.cs:1616]: Agreed that BIT is the proper type here.

Comment (saranshbhardwaj) [reply iAM-legacy/RoadCare3/FormAnalysis.cs:1616]: Okay. I was thinking that USE_REASON is a new field added, and added as a string type rather than bit type.

**Comment by gregglarson** [comment]
This change is to address a direct concern of Jake Walter about the _inadequate budget_ message in the Reason Table not giving enough information.

File: iAM-legacy/RoadCareSecurityOperations/SecOps.cs [hunk 2]
```diff
@@ -74,7 +74,10 @@ public bool ReadOnly
 
 		private AccessLevel GetPermissions( Dictionary<String, String> actionSpecifier )
 		{
-			AccessLevel grantedLevel = AccessLevel.Default;
+            return AccessLevel.CreateDestroy;
```
Comment (ARA-Trans) [inline iAM-legacy/RoadCareSecurityOperations/SecOps.cs:77]: Why was this return added here? The rest of the function now does nothing. What is the purpose of the change to this function?

File: iAM-legacy/Simulation/AnalysisMethod.cs [hunk 3]
```diff
@@ -19,11 +19,12 @@ public class AnalysisMethod
         private bool _isOMSUnlimited;
         private bool _isOMSTargetEnforced = true;
         public bool IsConditionalRSL { get; set; }
-
+        public bool IsUseReasons { get; set; }
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/AnalysisMethod.cs:22]: The name of this should be UsesReasons or similar.

File: iAM-legacy/Simulation/Investments.cs [hunk 4]
```diff
@@ -234,15 +234,17 @@ public void LoadBudgets()
         /// <param name="priority"></param>
         /// <param name="limits">Limits where treatments are split over multiple years</param>
         /// <returns></returns>
-        public string IsBudgetAvailable(float fAmount, String strBudget, String strYear, Hashtable hashAttributeValue,Priorities priority, List<ISplitTreatmentLimit> limits, string budgetLimitType, out string budgetHash, out ISplitTreatmentLimit splitTreatmentLimit)
+        public string IsBudgetAvailable(float fAmount, string strBudget, String strYear, Hashtable hashAttributeValue,Priorities priority, List<ISplitTreatmentLimit> limits, string budgetLimitType, out string budgetHash, out ISplitTreatmentLimit splitTreatmentLimit, out string reasonNoBudget)
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/Investments.cs:237]: A budget type class should be created to handle budget state and behaviors. Please create the appropriate object to handle this data and use it here.

File: iAM-legacy/Simulation/Investments.cs [hunk 5]
```diff
@@ -309,7 +311,10 @@ public string IsBudgetAvailable(float fAmount, String strBudget, String strYear,
                 }
             }
 
-
+            if(budgets.Count == 0)
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/Investments.cs:314]: These state checks should be moved to the new budget class.

File: iAM-legacy/Simulation/Investments.cs [hunk 6]
```diff
@@ -369,14 +374,16 @@ public string IsBudgetAvailable(float fAmount, String strBudget, String strYear,
                     var currentYear = year + i;
 
                     //Priority must be all years, match the current year or the current year must be greater than maximum year and the Maximum year must be a priority.
-                    if(!(priority.IsAllYears || priority.Years == currentYear || (currentYear > MaximumYear && priority.Years==MaximumYear)))
+                    if (!(priority.IsAllYears || priority.Years == currentYear || (currentYear > MaximumYear && priority.Years == MaximumYear)))
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/Investments.cs:377]: Priority domain logic belongs with priority objects and its data should be checked there, not in Investments.

File: iAM-legacy/Simulation/Simulation.cs [hunk 7]
```diff
@@ -663,7 +663,7 @@ private bool CompileAreaEquation()
         public bool GetSimulationMethod()
         {
             //Replace with Method information.
-            String strSelect = "SELECT ANALYSIS, BUDGET_CONSTRAINT,BENEFIT_VARIABLE, BENEFIT_LIMIT,RUN_TIME,USE_CUMULATIVE_COST,USE_ACROSS_BUDGET FROM " + cgOMS.Prefix + "SIMULATIONS WHERE SIMULATIONID='" + m_strSimulationID + "'";
+            String strSelect = "SELECT ANALYSIS, BUDGET_CONSTRAINT,BENEFIT_VARIABLE, BENEFIT_LIMIT,RUN_TIME,USE_CUMULATIVE_COST,USE_ACROSS_BUDGET,USE_REASONS FROM " + cgOMS.Prefix + "SIMULATIONS WHERE SIMULATIONID='" + m_strSimulationID + "'";
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/Simulation.cs:666]: cgOMS was removed from the system correct?

File: iAM-legacy/Simulation/Simulation.cs [hunk 8]
```diff
@@ -1303,7 +1304,7 @@ public bool CreateSimulationTable(String strNetworkID, String strSimulationID)
 
             var numberColumns = m_listAttributes.Count * (Investment.AnalysisPeriod + 1);
 
-            var numberTables = Math.Ceiling(Convert.ToDouble(numberColumns) / 900);
+            var numberTables = Math.Ceiling(Convert.ToDouble(numberColumns) / 750);
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/Simulation.cs:1307]: This magic number should be converted to a CONST which defines itself.

MAX_NUMBER_OF_MSSQL_COLUMNS or similar.

File: iAM-legacy/Simulation/Simulation.cs [hunk 9]
```diff
@@ -5379,7 +5381,7 @@ private void SpendUntilTargetsDeficientMetRoadCare(int nYear)
                     {
                         if (strBudget == null)
                         {
-                            strBudget = Investment.IsBudgetAvailable(fAmount, strMultipleBudget, nYear.ToString(), section.m_hashNextAttributeValue, priority,limits, "", out budgetHash,out limit);
+                            strBudget = Investment.IsBudgetAvailable(fAmount, strMultipleBudget, nYear.ToString(), section.m_hashNextAttributeValue, priority,limits, "", out budgetHash,out limit,out reasonNoBudget);
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/Simulation.cs:5384]: Another place where a budget object would be handy.
